File: workers/java/io/temporal/omes/ActivityDispatch.java
package io.temporal.omes;

import java.util.ArrayList;
import java.util.List;

// Maps an ExecuteActivityAction to its registered activity name and args.
// Shared by the workflow-scheduled path and the standalone-activity path.
public final class ActivityDispatch {
  final String type;
  final List<Object> args;

  private ActivityDispatch(String type, List<Object> args) {
    this.type = type;
    this.args = args;
  }

  static ActivityDispatch nameAndArgs(KitchenSink.ExecuteActivityAction act) {
    String type;
    List<Object> args = new ArrayList<>();
    if (act.hasDelay()) {
      type = "delay";
      args.add(act.getDelay());
    } else if (act.hasPayload()) {
      type = "payload";
      KitchenSink.ExecuteActivityAction.PayloadActivity payload = act.getPayload();
      byte[] inputData = new byte[payload.getBytesToReceive()];
      for (int i = 0; i < inputData.length; i++) {
        inputData[i] = (byte) (i % 256);
      }
      args.add(inputData);
      args.add(payload.getBytesToReturn());
    } else if (act.hasClient()) {
      type = "client";
      args.add(act.getClient());
    } else if (act.hasRetryableError()) {
      type = "retryable_error";
      args.add(act.getRetryableError());
    } else if (act.hasTimeout()) {
      type = "timeout";
      args.add(act.getTimeout());
    } else if (act.hasHeartbeat()) {
      type = "heartbeat";
      args.add(act.getHeartbeat());
    } else {
      type = "noop";
    }
    return new ActivityDispatch(type, args);
  }
}

File: workers/python/activity_dispatch.py
from datetime import timedelta
from typing import Any, Optional

from protos.kitchen_sink_pb2 import ExecuteActivityAction


def timeout_or_none(act: ExecuteActivityAction, field: str) -> Optional[timedelta]:
    if act.HasField(field):
        d = getattr(act, field)
        return timedelta(seconds=d.seconds, microseconds=d.nanos / 1000)
    return None


def activity_name_and_args(act: ExecuteActivityAction) -> tuple[str, list[Any]]:
    """Map an ExecuteActivityAction to its registered activity name and args.

    Shared by the workflow-scheduled path and the standalone-activity path.
    """
    if act.HasField("delay"):
        return "delay", [act.delay]
    elif act.HasField("payload"):
        input_data = bytes(i % 256 for i in range(act.payload.bytes_to_receive))
        return "payload", [input_data, act.payload.bytes_to_return]
    elif act.HasField("client"):
        return "client", [act.client]
    elif act.HasField("retryable_error"):
        return "retryable_error", [act.retryable_error]
    elif act.HasField("timeout"):
        return "timeout", [act.timeout]
    elif act.HasField("heartbeat"):
        return "heartbeat", [act.heartbeat]
    return "noop", []
